### DISCUSSION TITLE
shrpi sleep not waking up RPI5

### DISCUSSION THREAD

**Discussion by Shortyuk** [hatlabs/discussions#130] [Q&A]
Hi, I have recently bought a Sailor-Hat and added it to my RPI5. The SH seems to be workign as intended when I cut power to the system in that it does an automatic shutdown as I had hoped.

I have just added a CR1220 battery to allow the onboard RTC to work and I have enabled the RTC in the config.sys file while also disabling the RPI5s onboard RTC.

When I use the command shrpi sleep 120 the RPI shutdown but does not restart. I am then left in the position of struggling to restart the RPI as I have to disconnect the power and wait for the caps to discharge before I can add 12v power again and re power up the RPI. It takes an age for the caps to discharge.

So can anyone suggest what's wrong with the shrpi sleep 120 command not working and also is there a quick way without dismantling my enclosure and taking the SH off the RPI to discharge the caps or reboot the RPI when it gets stuck like this ?

**Comment by mairas** [comment]
Sorry for not noticing this question earlier!

Sleep functionality should definitely work with a Pi 5 as well. There's no difference in the behavior in that regard. I tried the functionality myself and found out the alarm time setting doesn't work properly on the latest Raspberry Pi OS version:

```
$ sudo rtcwake -m no -t 1727086500
rtcwake: assuming RTC uses UTC ...
rtcwake: set rtc wake alarm failed: Invalid argument
```

I'll investigate; it's probably an easy fix.

When testing, if you don't want to wait for the supercaps to deplete, you can briefly short pins 6 and 8 on the programming header to force the microcontroller to reset (the encircled pins in the image below). It's a bit of a hack though - I'll be releasing a new revision shortly that has a separate header for external inputs.

![shrpi_v2 0 0_top_render_ortho_reset](https://github.com/user-attachments/assets/9d53e679-ea6e-4cab-8e41-6a5c0ad74647)

**Reply by mairas** [reply]
It wasn't an easy fix. The wakeup feature for this particular RTC chip was disabled in Linux kernel 6.6. Other people need it too, and the fix is already present in the upstream kernel, but it might take some time until the Pi kernel has it. See the links in this issue: https://github.com/hatlabs/SH-RPi-daemon/issues/151

**Comment by Shortyuk** [comment]
Hi Matt,

Thanks for looking into this and the top on shorting the jumpers. I'll keep
an eye out for the PI Kernel being patched.

**Reply by mairas** [reply]
No worries. It was a bit of a rabbit hole, I have to admit. :-)

One thing I didn't mention: it is possible to bypass the kernel altogether and communicate directly with the RTC chip using `i2cget` 
 and `i2cset` (or the Python libraries). The RTC I2C address is 0x51. See the data sheet for the registers: https://www.nxp.com/docs/en/data-sheet/PCF8563.pdf

For example, you can read the time and date registers as follows:

```
sudo i2cget -f -y 1 0x51 0x02 i 7
```

(Read from bus 1, address 0x51, 7 bytes from register 0x02 onwards.)

Force flag is needed because the kernel RTC driver reserves the I2C address.

Also note that the registers are encoded as [BCD](https://en.wikipedia.org/wiki/Binary-coded_decimal).

**Reply by wkuna** [reply]
I wonder if this issue got resolved in the end. I'm on 6.12.34+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.12.34-1+rpt1~bookworm (2025-06-26) aarch64, and I still can't get the sleep functionality to work

**Reply by rowanbsmith** [reply]
You need to add wake-source to the dtoverlay on 6.12 - see: https://github.com/hatlabs/SH-RPi-daemon/issues/151#issuecomment-4375280970
